### DISCUSSION TITLE
Can Dexed receive voices via sysex?

### DISCUSSION THREAD

**Discussion by probonopd** [asb2m10/dexed#352] [Q&A]
Can Dexed receive voices via sysex?

My use case:
Using Dexed as a patch editor for [MiniDexed](https://github.com/probonopd/MiniDexed) (Dexed running as 'embedded firmware' on Raspberry Pi). It would be nice if Dexed could show the knob settings depending on the voice selected on MiniDexed. For this purpose, MiniDexed sends the currently selected voice as a sysex.

Another use case:
I would like to put voice data into my MIDI tracks in my DAW (e.g., Reaper). When playing those tracks, the voice data from the sysex should be loaded into the Dexed VST.

**Comment by asb2m10** [comment]
Can you elaborate this ? If MiniDexed sends the selected voice on sysex, dexed parms should be updated accordingly. You mean this is not the case ?

**Reply by probonopd** [reply]
> If MiniDexed sends the selected voice on sysex, dexed parms should be updated accordingly. 

Thanks @asb2m10, this answers my question.

> dexed parms should be updated accordingly

In my last tests, this did not work but I was not sure whether it _should_ work. Now that I know, I will search for the bug in MiniDexed ;-)

Whenever you switch voices (and on power-on), MiniDexed should send the voice data sysex.

**Reply by asb2m10** [reply]
See implementation: https://github.com/asb2m10/dexed/blob/bda3dc2aebfe528d43e66ac7bfef8f828daca9f4/Source/PluginProcessor.cpp#L595

**Comment by probonopd** [comment]
Can't seem to get it to work... this time, I am trying to load a voice from a sysex message on a MIDI track in Reaper. Here is my Reaper file:

[LoadingVoiceDataDoesNotWork.zip](https://github.com/probonopd/MiniDexed/files/13627469/LoadingVoiceDataDoesNotWork.zip)

To see the sysex, double click the item e.g., on the first track, then click the third icon from the left in the toolbar to see the MIDI messages. Double click the sysex one to see its contents.

Playing this file will play the notes, but will not cause the voice contained in the sysex to be loaded into Dexed.

Do you have an idea why this is not working?

**Comment by probonopd** [comment]
I verified that Dexed sends voice data in the following format:

`F0430000011B140A1C185E5340002A3600010000000340000300071F08171C635742002C050E03000000025C000100074A300C24516344002F080D00000000056300010007182A0D0A635C42002D090C00000000003D000200071E2E080C636343002D080C010000000054000100074C12042C546346002C0A0D0000000005630001000E5E435F3C323232320206013844003F010301185341484152412020202078F7` (163 bytes)

To get this string on Linux: 
* Create a virtual MIDI port like this: `amidi -p virtual -d` in a console
* In Reaper, create a track and put Dexed on it
* In Dexed PARMS, select "Virtual RawMIDI" as the "DX7 Out"
* Using Dexed, send a voice to the DX7 (this sends it to the virtual MIDI port)
* Now you can see the string in the terminal window

I also verified that Dexed receives voice data in the same format.

To verify this:
* In Reaper, create 2 tracks and put one Dexed instance each on them
* Set first instance to use "Midi Through Port-0" as its "DX7 Out"
* Set second instance to use "Midi Through Port-0" as its "DX7 In"
* Send a voice fro the first instance to the DX7, it will send the voice data to the other instance successfully

__However,__ when I put the same sysex string (without the leading `F0` and the trailing `F7` which Reaper adds automatically) onto a MIDI track and play that with Dexed, it won't load the sysex from the playing MIDI track. __Why? Are sysex only accepted if they are coming from the "DX7 In" MIDI interface but are ignored when part of the MIDI track being played?__

**Reply by probonopd** [reply]
That seems to be the case.

The following works:
* In Reaper, create one more track
* On this track, put the sysex
* DON'T enable Dexed on this track
* Make this track send its MIDI output to "virtual" (created using `amidi -p virtual -d`)
* On the track on which Dexed is and on which the MIDI notes to be played are, set "DX7 In" to "Virtual RawMIDI"
* Play the tracks. When the sysex track is playing, the track containing Dexed and the MIDI notes will load the voice from the sysex

I think this proves that currently Dexed loads voice data only from the "DX7 In" MIDI device but not from its own MIDI track (where the notes to be played are in a DAW).

Could this be enabled please? Opened a feature request for it:

* https://github.com/asb2m10/dexed/issues/401

**Comment by probonopd** [comment]
Thank you very much, works for me in [482c8cd](https://github.com/asb2m10/dexed/commit/482c8cd791d55667bdc4aa48f5914f6e659ae8aa).
